### PR TITLE
Bridge destruction

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

Habitat as a whole has a fair number of bridge currently.  By my count, cpp sim -> bound python sim, bound python sim -> habitat-sim python simulator, habitat-sim python simulator -> habitat-api python simulator, habitat-api python simulator -> env. 

Bridges do have their use, they allow for progressively higher level APIs as you get closer and closer to a use-case.  The `geodesic_distance` function in the habitat-api python simulator is a good example of this.  They also allow us to use python in locations where maximal performance doesn't matter and python is a better tool for the job than cpp.

Where the bridges fall down is that if you want to have a method in the cpp simulator and get it all the way up to the highest level, you have to write a lot of wrapper functions or do `_sim._sim._sim`.  This makes development and iteration on new things annoying.

This PR is a proposal for building this functionality with inheritance instead of using private member and wrapper method.  I rebuilt the of the `bound python sim -> habitat-sim python simulator` bridge with this idea.  This specific bridge does have one other use-case, the ability to destroy all things related to OpenGL explicitly.  I have replicated this functionality with a very carefully written `close()` method.  

Here are the pros and cons the way I see it:

### Pros
- Less boiler plate code (I deleted almost 200 lines of code!)
- Less things to document
- It is currently confusing why once a method is bound to python from cpp why it doesn't show up in the simulator class

### Cons
- That `close()` method in cpp may or may not be a landmine.  We don't add new top level things often, but when we do, it's an extra step that needs to be done and done right
- Less delineation between public/private APIs

## How Has This Been Tested

Tests pass.

## Types of changes

Fairly heavy refactoring.

### CODE REVIEW

**Comment by Skylion007** [comment]
More generally, is there any reason we have so many get/set functions exposed in Python instead of using PyBind properties to do it in a more pythonic way. Example from the PyBind tutorial:
```
 .def_property("name", &Pet::getName, &Pet::setName)
```

**Comment by erikwijmans** [comment]
> If we are going be changing the Python side of the API

This is not changing the python side of the API.  It is just removing the need for a lot wrapper functions.

**Comment by aclegg3** [comment]
Thanks, Erik. I like this refactor. Nice to see that one extra layer of simulator.py functions removed.

File: src/esp/bindings/SimBindings.cpp [hunk 1]
```diff
@@ -133,7 +146,7 @@ void initSimBindings(py::module& m) {
       .def("set_object_bb_draw", &Simulator::setObjectBBDraw, "draw_bb"_a,
            "object_id"_a, "sceneID"_a = 0)
       .def("recompute_navmesh", &Simulator::recomputeNavMesh, "pathfinder"_a,
-           "navmesh_settings"_a, "include_static_objects"_a)
+           "navmesh_settings"_a, "include_static_objects"_a = true)
```
Comment (mathfac) [inline src/esp/bindings/SimBindings.cpp:149]: It was `false originally`
```suggestion
           "navmesh_settings"_a, "include_static_objects"_a = false)
```

File: src/esp/sim/Simulator.h [hunk 2]
```diff
@@ -65,6 +65,14 @@ class Simulator {
   explicit Simulator(const SimulatorConfiguration& cfg);
   virtual ~Simulator();
 
+  /**
+   * @brief Closes the simulator and frees all loaded assets and GPU contexts
+   *
+   * @warning Must reset the simulator to its "just after constructor" state for
+   * python inheritance to function correctly
```
Comment (mathfac) [inline src/esp/sim/Simulator.h:72]: Can we add more context here, what should be added here: each shared pointer member or etc and what will happen if that won't be done.

File: src/esp/sim/Simulator.cpp [hunk 3]
```diff
@@ -36,9 +36,37 @@ Simulator::Simulator(const SimulatorConfiguration& cfg) {
 
 Simulator::~Simulator() {
   LOG(INFO) << "Deconstructing Simulator";
+  close();
+}
+
+void Simulator::close() {
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply src/esp/sim/Simulator.cpp:42]: These are still all smart pointers so nothing will leak, what you will get is weird behavior when doing `close() -> reconfigure()` however.

File: habitat_sim/simulator.py [hunk 4]
```diff
@@ -43,7 +43,7 @@ class Configuration(object):
 
 
 @attr.s(auto_attribs=True)
-class Simulator:
+class Simulator(SimulatorBackend):
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_sim/simulator.py:46]: No idea what you are referring to here.

**Comment by aclegg3** [comment]
> More generally, is there any reason we have so many get/set functions exposed in Python instead of using PyBind properties to do it in a more pythonic way. 

Thanks for this suggestion. I'll add this to the list of considerations for python API design. This can be a separate change.

**Comment by erikwijmans** [comment]
What are folks thoughts about merging this before or after the release cut for v0.1.5?

**Pull request merged**